components/browser/thumbnails/src/test/java/mozilla/components/browser/thumbnails/BrowserThumbnailsTest.kt

```diff
@@ -4,98 +4,104 @@
+    private val tabId = "test-tab"
 
     @Before
     fun setup() {
-        val engine = mock<Engine>()
-        mockSessionManager = spy(SessionManager(engine))
-        mockEngineView = mock()
-        thumbnails = BrowserThumbnails(testContext, mockEngineView, mockSessionManager)
+        Dispatchers.setMain(testDispatcher)
+        store = spy(BrowserStore(BrowserState(
+            tabs = listOf(
+                createTab("https://www.mozilla.org", id = tabId)
+            ),
+            selectedTabId = tabId
+        )))
+        engineView = mock()
+        thumbnails = BrowserThumbnails(testContext, engineView, store)
+    }
+
+    @After
+    fun tearDown() {
+        Dispatchers.resetMain()
+        testDispatcher.cleanupTestCoroutines()
     }
 
     @Test
-    fun `when feature is stop must not capture thumbnail when a site finish loading`() {
+    fun `do not capture thumbnail when feature is stopped and a site finishes loading`() {
         thumbnails.start()
         thumbnails.stop()
 
-        val session = getSelectedSession()
+        // Initial loading state is false
+        verify(engineView).captureThumbnail(any())
 
-        session.notifyObservers {
-            onLoadingStateChanged(session, false)
-        }
+        store.dispatch(ContentAction.UpdateThumbnailAction(tabId, mock())).joinBlocking()
 
-        verify(mockEngineView, never()).captureThumbnail(any())
+        verifyNoMoreInteractions(engineView)
     }
 
+    @Suppress("UNCHECKED_CAST")
     @Test
-    fun `feature must capture thumbnail when a site finish loading`() {
-        thumbnails.start()
+    fun `feature must capture thumbnail when a site finishes loading`() {
+        val bitmap: Bitmap? = mock()
 
-        val session = getSelectedSession()
-
-        session.notifyObservers {
-            onLoadingStateChanged(session, false)
-        }
-
-        verify(mockEngineView).captureThumbnail(any())
-    }
+        store.dispatch(ContentAction.UpdateLoadingStateAction(tabId, true)).joinBlocking()
 
-    @Test
-    fun `feature must capture thumbnail when title changes`() {
         thumbnails.start()
 
-        val session = getSelectedSession()
+        `when`(engineView.captureThumbnail {})
+            .thenAnswer { // if engineView responds with a null
+                (it.arguments[0] as (Bitmap?) -> Unit).invoke(null)
+            }.thenAnswer { // if engineView responds with a bitmap
+                (it.arguments[0] as (Bitmap?) -> Unit).invoke(bitmap)
+            }
+
+        verify(store, never()).dispatch(ContentAction.UpdateThumbnailAction(tabId, bitmap!!))
 
-        session.notifyObservers {
-            onTitleChanged(session, "test")
-        }
+        store.dispatch(ContentAction.UpdateLoadingStateAction(tabId, false)).joinBlocking()
 
-        verify(mockEngineView).captureThumbnail(any())
+        verify(store, never()).dispatch(ContentAction.UpdateThumbnailAction(tabId, bitmap))
```

Comment:
Here you'd want to verify that you dispatched the `UpdateThumbnailAction`. So, this shouldn't have `never()` at least the way I read it.

Reply:
So, basically, something like this?

```kotlin
@Test
    fun `feature must capture thumbnail when a site finishes loading`() {
        val bitmap: Bitmap? = mock()

        store.dispatch(ContentAction.UpdateLoadingStateAction(tabId, true)).joinBlocking()

        thumbnails.start()

        `when`(engineView.captureThumbnail(any()))
            .thenAnswer { // if engineView responds with a bitmap
                (it.arguments[0] as (Bitmap?) -> Unit).invoke(bitmap)
            }

        verify(store, never()).dispatch(ContentAction.UpdateThumbnailAction(tabId, bitmap!!))

        store.dispatch(ContentAction.UpdateLoadingStateAction(tabId, false)).joinBlocking()

        verify(store).dispatch(ContentAction.UpdateThumbnailAction(tabId, bitmap))
    }
```

Reply:
Thanks for catching! I've fixed this test and wrote some more to cover the other null cases too.